observability/src/main/java/io/grpc/observability/metadata/MetadataConfig.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability.metadata;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.api.client.http.HttpTransport;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.Internal;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Class to read Google Metadata Server values. */
+@Internal
+public final class MetadataConfig {
+  private static final Logger logger = Logger.getLogger(MetadataConfig.class.getName());
+
+  private static final int TIMEOUT_MS = 5000;
+  private static final String METADATA_URL = "http://metadata.google.internal/computeMetadata/v1/";
+  HttpRequestFactory requestFactory;
+
+  @VisibleForTesting public MetadataConfig(HttpTransportFactory transportFactory) {
+    HttpTransport httpTransport = transportFactory.create();
+    requestFactory = httpTransport.createRequestFactory();
+  }
+
+  /** gets all the values from the MDS we need to set in our logging tags. */
+  public Map<String, String> getAllValues() {
+    HashMap<String, String> map = new HashMap<>();
+    //addValueFor(map, "instance/hostname", "GCE_INSTANCE_HOSTNAME");
+    addValueFor(map, "instance/id", "gke_node_id");
+    //addValueFor(map, "instance/zone", "GCE_INSTANCE_ZONE");
+    addValueFor(map, "project/project-id", "project_id");
+    addValueFor(map, "project/numeric-project-id", "project_numeric_id");
+    addValueFor(map, "instance/attributes/cluster-name", "cluster_name");
+    addValueFor(map, "instance/attributes/cluster-uid", "cluster_uid");
+    addValueFor(map, "instance/attributes/cluster-location", "location");
+    try {
+      requestFactory.getTransport().shutdown();
```

Comment:
Good catch! I did something better (I think). I created an `init()` method. So one can call `init()` and `getAllValues()` any number of times (although it is needed only once in the beginning).

Reply:
So now I can create a MetadataConfig instance, forget to call init(), call getAllValues() and then get a NPE, right? Since creating the RequestFactory is not expensive and getAllValues() is only expected to be called once, I would create it there. To me the init() method creates extra complication - but I'll let you make the call 😄

Reply:
From the (internal) go/java-practices/constructors#avoid-doing-work-in-constructors doc: 

"The simplest and commonest kind of constructor just assigns each of its parameters to a field of the same name and type, perhaps after validation.... The more additional work the constructor does, the harder it will be to test the class, and especially to test other classes that need it."